File: examples/simple_websocket/test_client/groups/ChatGroup.js


class ChatGroup {
  constructor(client) {
    this.client = client;

    this.username = null;

    this.$msg = document.getElementById('chat-messages');
  }

  request_register(username) {

    this.client.request('Chat:register', {
      username: username,
    }).then((resp) => {

      this.username = username;

      for (let msg of resp.history) {
        this.$msg.innerHTML += "<br/>"+msg;
      }
    });
  }

  new_user({username}) {
    this.$msg.innerHTML += "<br/>New user: <b>"+username+"</b>";
  }

  request_message(msg) {

    this.client.request('Chat:message', {
      msg: msg,
    }).then(function() {

    });
  }

  new_message({username, msg}) {
    this.$msg.innerHTML += "<br/>"+username+": "+msg;
  }
}

File: examples/simple_websocket/wsapp/groups/ChatGroup.py
from eme.websocket import WebsocketApp, WSClient

from wsapp.entities import User


class ChatGroup:

    def __init__(self, server):
        self.server: WebsocketApp = server
        self.group = 'Chat'

        self.history = []

        # This is another way to configure routes that do not require authentication
        self.server.no_auth.update([
            'Chat:register'
        ])

    def register(self, username: str, client: WSClient):
        # create new user object (normally we should authenticate with oauth2 or other means!)
        client.user = User(username=username, client=client)

        # by adding user attribute, the client will be able to access routes that require authentication

        self.server.broadcast({
            "route": "Chat:new_user",
            "username": username
        })

        # send history only to the new client:
        # you can specify route here too, but if you don't the same route will be used
        return {
            "username": username,
            "history": self.history
        }

    def message(self, msg: str, user: User):
        self.history.append(user.username + ': ' + msg)

        self.server.broadcast({
            "route": "Chat:new_message",
            "username": user.username,
            "msg": msg
        })
